service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,328 @@
+		if err := activity.GetHeartbeatDetails(ctx, &hbd); err == nil {
+			startOver = false
+		} else {
+			//TODO error metrics
+			getActivityLogger(ctx).Error("Failed to recover from last heartbeat, start over from beginning", tag.Error(err))
+		}
+	}
+
+	if startOver {
+		resp, err := batcher.svcClient.CountWorkflowExecutions(ctx, &shared.CountWorkflowExecutionsRequest{
+			Domain: common.StringPtr(batchParams.DomainName),
+			Query:  common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+		hbd.TotalEstimate = resp.GetCount()
+	}
+	rateLimiter := rate.NewLimiter(rate.Limit(batchParams.RPS), batchParams.RPS)
+	taskCh := make(chan taskDetail, pageSize)
+	var succCount, errCount int32
+	for i := 0; i < batchParams.Concurrency; i++ {
+		go startTaskProcessor(ctx, batchParams, taskCh, rateLimiter, &succCount, &errCount)
+	}
+
+	for {
+		// TODO https://github.com/uber/cadence/issues/2154
+		//  Need to improve scan concurrency because it will hold a ES resource for a long time.
+		//  we can't use list API because terminate / reset will mutate the result.
+		resp, err := batcher.svcClient.ScanWorkflowExecutions(ctx, &shared.ListWorkflowExecutionsRequest{
+			PageSize:      common.Int32Ptr(int32(pageSize)),
+			Domain:        common.StringPtr(batchParams.DomainName),
+			NextPageToken: hbd.PageToken,
+			Query:         common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+
+		// clear all counters
+		atomic.StoreInt32(&succCount, 0)
+		atomic.StoreInt32(&errCount, 0)
+
+		// send all tasks
+		for _, wf := range resp.Executions {
+			taskCh <- taskDetail{
+				execution: *wf.Execution,
+				attempts:  0,
+				hbd:       hbd,
+			}
+		}
+		batchCount := len(resp.Executions)
+
+		// wait for counters indicate this batch is done
+	Loop:
+		for {
+			select {
+			case <-time.Tick(time.Second):
+				if int32(batchCount) == atomic.LoadInt32(&succCount)+atomic.LoadInt32(&errCount) {
+					break Loop
+				}
+			case <-ctx.Done():
+				return nil
+			}
+		}
+
+		hbd.CurrentPage++
+		hbd.PageToken = resp.NextPageToken
+		hbd.SuccessCount += atomic.LoadInt32(&succCount)
+		hbd.ErrorCount += atomic.LoadInt32(&errCount)
+		activity.RecordHeartbeat(ctx, hbd)
+
+		if len(hbd.PageToken) == 0 {
+			break
+		}
+	}
+
+	return nil
+}
+
+func startTaskProcessor(
+	ctx context.Context, batchParams BatchParams, taskCh chan taskDetail,
+	limiter *rate.Limiter, doneCount, errCount *int32) {
+
+	for {
+		select {
+		case task := <-taskCh:
+			if isDone(ctx) {
+				return
+			}
+			var err error
+
+			switch batchParams.BatchType {
+			case BatchTypeTerminate:
+				err = processOneTerminateTask(ctx, limiter, task, batchParams)
+			}
+			if err != nil {
+				getActivityLogger(ctx).Error("Failed to process task", tag.Error(err))
+				// put back to the channel if less than attemptsOnError
+				task.attempts++
+				if task.attempts >= batchParams.AttemptsOnRetryableError {
```

Comment:
So expectation is that `err` if returned has to be retryable? That seems prone to errors.. The error handling here should deal with different types of errors - we shouldn't expect the specific methods to handle the retryable logic.